bsi/src/main/java/org/roaringbitmap/bsi/longlong/Roaring64NavigableMapSliceIndex.java

```diff
@@ -0,0 +1,622 @@
+      }
+    }
+
+    return valueAt(minValuesId.first());
+  }
+
+  private long maxValue() {
+    if (ebM.isEmpty()) {
+      return 0;
+    }
+
+    Roaring64NavigableMap maxValuesId = ebM;
+    for (int i = bA.length - 1; i >= 0; i -= 1) {
+      Roaring64NavigableMap tmp = Roaring64NavigableMap.and(maxValuesId, bA[i]);
+      if (!tmp.isEmpty()) {
+        maxValuesId = tmp;
+      }
+    }
+
+    return valueAt(maxValuesId.first());
+  }
+
+  private long valueAt(long columnId) {
+    long value = 0;
+    for (int i = 0; i < this.bitCount(); i += 1) {
+      if (this.bA[i].contains(columnId)) {
+        value |= (1L << i);
+      }
+    }
+
+    return value;
+  }
+
+  /**
+   * RunOptimize attempts to further compress the runs of consecutive values found in the bitmap
+   */
+  public void runOptimize() {
+    this.ebM.runOptimize();
+
+    for (Roaring64NavigableMap integers : this.bA) {
+      integers.runOptimize();
+    }
+    this.runOptimized = true;
+  }
+
+  /**
+   * hasRunCompression returns true if the bitmap benefits from run compression
+   */
+
+  public boolean hasRunCompression() {
+    return this.runOptimized;
+  }
+
+  /**
+   * GetExistenceBitmap returns a pointer to the underlying existence bitmap of the BSI
+   */
+
+  public Roaring64NavigableMap getExistenceBitmap() {
+    return this.ebM;
+  }
+
+  public int bitCount() {
+    return this.bA.length;
+  }
+
+  public long getLongCardinality() {
+    return this.ebM.getLongCardinality();
+  }
+
+  /**
+   * GetValue gets the value at the column ID.  Second param will be false for non-existence values.
+   */
+  public Pair<Long, Boolean> getValue(long columnId) {
+    boolean exists = this.ebM.contains(columnId);
+    if (!exists) {
+      return Pair.newPair(0L, false);
+    }
+
+    return Pair.newPair(valueAt(columnId), true);
+  }
+
+
+  private void clear() {
+    this.maxValue = 0;
+    this.minValue = 0;
+    this.ebM = null;
+    this.bA = null;
+  }
+
+  /**
+   * Serialize this bitmapsliceindex(bsi).
+   * <p>
+   * There is no specification for now: it may change from one java version to
+   * another. This serialization format is compatible with bsi of roaring(Go)
+   * <p>
+   * Consider calling {@link #runOptimize} before serialization to improve compression.
+   * <p>
+   * The current bsi is not modified.
+   *
+   * @param output the DataOutput stream
+   * @throws IOException Signals that an I/O exception has occurred.
+   */
+  public void serialize(DataOutput output) throws IOException {
+    this.ebM.serialize(output);
+    for (Roaring64NavigableMap rb : this.bA) {
+      rb.serialize(output);
+    }
+  }
+
+  public void serialize(ByteBuffer buffer) throws IOException {
+    this.ebM.serialize(buffer);
+    for (Roaring64NavigableMap rb : this.bA) {
+      rb.serialize(buffer);
+    }
+  }
+
+  /**
+   * Deserialize (retrieve) this bitmapsliceindex(bsi).
+   * <p>
+   * There is no specification for now: it may change from one java version to
```

Comment:
I felt `having the same serialization format than BSI-64 Go implementation` some sort of spec by itself. Could you add some comments about the current format?

Reply:
i think current BSI serialization format is not so elegant in Go implementation.
it simply formats ebM and bA[], w/o any meta/checksum info. And this is not friendly to deserialization. @lemire @blacelle 
So this implementation just follow Go's way. I will add some more comments about the serialization format.